bus-mapping/src/evm/opcodes/return.rs

```diff
@@ -1,175 +1,199 @@
-use crate::circuit_input_builder::{CircuitInputStateRef, ExecStep};
-use crate::evm::Opcode;
-use crate::Error;
-use eth_types::GethExecStep;
+use super::Opcode;
+use crate::circuit_input_builder::{CopyDataType, CopyEvent, NumberOrHash};
+use crate::operation::MemoryOp;
+use crate::{
+    circuit_input_builder::CircuitInputStateRef,
+    evm::opcodes::ExecStep,
+    operation::{CallContextField, RW},
+    Error,
+};
+use eth_types::{Bytecode, GethExecStep, ToWord, H256};
+use ethers_core::utils::keccak256;
 
 #[derive(Debug, Copy, Clone)]
 pub(crate) struct Return;
 
+// TODO: rename to indicate this handles REVERT (and maybe STOP)?
 impl Opcode for Return {
     fn gen_associated_ops(
         state: &mut CircuitInputStateRef,
-        geth_steps: &[GethExecStep],
+        steps: &[GethExecStep],
     ) -> Result<Vec<ExecStep>, Error> {
-        let geth_step = &geth_steps[0];
-        let exec_step = state.new_step(geth_step)?;
+        let step = &steps[0];
+        let mut exec_step = state.new_step(step)?;
 
-        let current_call = state.call()?.clone();
-        let offset = geth_step.stack.nth_last(0)?.as_usize();
-        let length = geth_step.stack.nth_last(1)?.as_usize();
+        let offset = step.stack.nth_last(0)?;
+        let length = step.stack.nth_last(1)?;
+        state.stack_read(&mut exec_step, step.stack.nth_last_filled(0), offset)?;
+        state.stack_read(&mut exec_step, step.stack.nth_last_filled(1), length)?;
+
+        if !length.is_zero() {
+            state
+                .call_ctx_mut()?
+                .memory
+                .extend_at_least((offset.low_u64() + length.low_u64()).try_into().unwrap());
+        }
+
+        let call = state.call()?.clone();
+        let is_root = call.is_root;
+        for (field, value) in [
+            (CallContextField::IsRoot, is_root.to_word()),
+            (CallContextField::IsCreate, call.is_create().to_word()),
+            (CallContextField::IsSuccess, call.is_success.to_word()),
+            (CallContextField::CallerId, call.caller_id.into()),
+            (
+                CallContextField::ReturnDataOffset,
```

Comment:
I think this implementation is not quite following the spec.  For example, the lookup for `ReturnDataOffset` and `ReturnDataLength` is done conditionally in the case `not instruction.curr.is_root and not instruction.curr.is_create`.

Also the cases implemented here don't follow exactly the ones in the spec, so the logic is a bit different.

Could you annotate the code to make it clear where the different paths `A, B, C, D` are implemented?  The same applies for the circuit code.

And if you made the decision to deviate from the spec, could you explain why?

Reply:
I have added references to the specs in the comments here and updated the specs in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/283 so it matches the implemenation here.